demos/python_demos/face_recognition_demo/face_recognition_demo.py

```diff
@@ -0,0 +1,1073 @@
+        self.output_shape = model.outputs[self.output_blob].shape
+
+        assert np.array_equal(self.ADAS_INPUT_SHAPE, self.input_shape) or \
+               np.array_equal(self.RETAIL_INPUT_SHAPE, self.input_shape), \
+            "Expected model input shape %s, but got %s" % \
+            (" or ".join([self.ADAS_INPUT_SHAPE, self.RETAIL_INPUT_SHAPE]),
+             self.input_shape)
+
+        assert len(self.output_shape) == 4 and \
+               self.output_shape[3] == self.Result.OUTPUT_SIZE, \
+            "Expected model output shape with %s outputs" % \
+            (self.Result.OUTPUT_SIZE)
+
+        assert 0.0 <= confidence_threshold and confidence_threshold <= 1.0, \
+            "Confidence threshold is expected to be in range [0; 1]"
+        self.confidence_threshold = confidence_threshold
+
+        assert 0.0 < roi_scale_factor, "Expected positive ROI scale factor"
+        self.roi_scale_factor = roi_scale_factor
+
+    def get_input_shape(self):
+        return self.input_shape
+
+    def preprocess(self, frame):
+        assert len(frame.shape) == 4, "Frame shape should be [1, c, h, w]"
+        input = self._resize(frame, self.input_shape)
+        return input
+
+    def start_async(self, frame):
+        input = self.preprocess(frame)
+        self.enqueue(input)
+
+    def enqueue(self, input):
+        return super(FaceDetector, self).enqueue({self.input_blob: input})
+
+    def get_roi_proposals(self, frame):
+        outputs = self.get_outputs()[0][self.output_blob]
+        # outputs shape is [N_requests, 1, 1, N_max_faces, 7]
+
+        frame_width = frame.shape[-1]
+        frame_height = frame.shape[-2]
+
+        results = []
+        for output in outputs[0][0]:
+            result = self.Result(output)
+            if result.confidence < self.confidence_threshold:
+                break # results are sorted by confidence decrease
+
+            self._clip(result, 1, 1)
+            self._resize_roi(result, frame_width, frame_height)
+            self._rescale_roi(result, self.roi_scale_factor)
+            self._clip(result, frame_width, frame_height)
+
+            results.append(result)
+
+        return results
+
+    def _rescale_roi(self, result, roi_scale_factor=1.0):
+        result.position -= result.size * 0.5 * (roi_scale_factor - 1.0)
+        result.size *= roi_scale_factor
+        return result
+
+    def _resize_roi(self, result, frame_width, frame_height):
+        result.position[0] *= frame_width
+        result.position[1] *= frame_height
+        result.size[0] = result.size[0] * frame_width - result.position[0]
+        result.size[1] = result.size[1] * frame_height - result.position[1]
+        return result
+
+    def _clip(self, result, width, height):
+        min = [0, 0]
+        max = [width, height]
+        result.position[:] = clip(result.position, min, max)
+        result.size[:] = clip(result.size, min, max)
+        return result
+
+
+class HeadPoseEstimator(Module):
```

Comment:
Do we really need headpose in this sample, too?

Reply:
Why not? I don't see any issues with this, also it is good for demonstration purposes as this model can be executed in parallel with others - nothing else depends on its outputs.

Reply:
It adds entropy for little value (in my opinion), that's all. We already have interactive_face_detection demo that's kinda went out of proportion. Anyway, I don't think it's a blocker, but I'd leave it out.